### PR TITLE
Istio1.15 support

### PR DESCRIPTION
**Pull request by luciferinlove**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | yes
| Deprecations?   | no
| Related tickets |
| License         | Apache 2.0

### What's in this PR?

Istio1.15 support

### Checklist

- [X] Implementation tested
- [X] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [X] Logging code meets the guideline
- [X] User guide and development docs updated (if needed)

### CODE REVIEW

File: api/v1alpha1/common.proto [hunk 1]
```diff
@@ -60,15 +57,15 @@ message ContainerImageConfiguration {
 
     // ImagePullSecrets is an optional list of references to secrets to use for pulling any of the images.
     // +optional
-    repeated k8s.io.api.core.v1.LocalObjectReference imagePullSecrets = 4 [(gogoproto.nullable) = false];
```
Comment (Laci21) [inline api/v1alpha1/common.proto:63]: Do we have the same functionality as `[(gogoproto.nullable) = false]`?

Comment (Laci21) [reply api/v1alpha1/common.proto:63]: Or for `[(gogoproto.wktpointer) = true]` or others?

File: api/v1alpha1/common.proto [hunk 2]
```diff
@@ -125,24 +122,24 @@ message BaseKubernetesResourceConfig {
     // +optional
     string priorityClassName = 10;
 
-    // google.protobuf.Int32Value replicaCount = 1 [(gogoproto.wktpointer) = true];
+    // google.protobuf.Int32Value replicaCount = 1;
```
Comment (Laci21) [inline api/v1alpha1/common.proto]: Not sure why this is commented out. If not needed we could delete this, otherwise we should use this field. I guess we don't need it, but did not check.

File: api/v1alpha1/istiocontrolplane.proto [hunk 3]
```diff
@@ -126,7 +125,7 @@ message IstioControlPlaneSpec {
 }
 
 enum ModeType {
-    UNSPECIFIED = 0;
+    ModeType_UNSPECIFIED = 0;
```
Comment (Laci21) [inline api/v1alpha1/istiocontrolplane.proto:128]: Is this still not working if you use the same name `UNSPECIFIED` multiple places? Other option, what you simply delete this field, would that work maybe?

File: api/v1alpha1/istiocontrolplane_types.go [hunk 4]
```diff
@@ -54,16 +53,20 @@ type IstioControlPlane struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   *IstioControlPlaneSpec  `json:"spec,omitempty"`
-	Status IstioControlPlaneStatus `json:"status,omitempty"`
+	Spec   *IstioControlPlaneSpec   `json:"spec,omitempty"`
+	Status *IstioControlPlaneStatus `json:"status,omitempty"`
```
Comment (Laci21) [inline api/v1alpha1/istiocontrolplane_types.go:57]: Why was this needed exactly?

File: deploy/charts/istio-operator/Chart.yaml [hunk 5]
```diff
@@ -9,7 +9,7 @@ sources:
 icon: https://istio.io/latest/img/istio-whitelogo-bluebackground-framed.svg
 
 # Based on support status of Istio releases: https://istio.io/latest/docs/releases/supported-releases/#support-status-of-istio-releases
-kubeVersion: ">= 1.20.0-0 <= 1.23.0-0"
+kubeVersion: ">= 1.21.0-0 <= 1.24.3-0"
```
Comment (Laci21) [inline deploy/charts/istio-operator/Chart.yaml]: this will probably be >= 1.22.0-0 <= 1.26.0-0

File: deploy/charts/istio-operator/Chart.yaml [hunk 6]
```diff
@@ -9,7 +9,7 @@ sources:
...
 
-version: 2.0.12
-appVersion: "v2.13.4"
+version: 2.0.13
+appVersion: "v2.14.0"
```
Comment (Laci21) [inline deploy/charts/istio-operator/Chart.yaml]: We usually release rc releases first for testing, but I'm not consisting if you can test otherwise as well easily, let's just not release the stable version until the code is reviewed/solution is tested.

File: deploy/charts/istio-operator/README.md [hunk 7]
```diff
@@ -1,11 +1,11 @@
 # Istio-operator chart
 
-[Istio-operator](https://github.com/banzaicloud/istio-operator/tree/release-1.13) is a Kubernetes operator to deploy and manage [Istio](https://istio.io/) resources for a Kubernetes cluster.
+[Istio-operator](https://github.com/banzaicloud/istio-operator/tree/release-1.15) is a Kubernetes operator to deploy and manage [Istio](https://istio.io/) resources for a Kubernetes cluster.
 
 ## Prerequisites
 
 - Helm3
-- Kubernetes 1.19.0 - 1.22.x
+- Kubernetes 1.21.0 - 1.24.x
```
Comment (Laci21) [inline deploy/charts/istio-operator/README.md:8]: same as above

File: deploy/charts/istio-operator/templates/operator-rbac.yaml [hunk 8]
```diff
@@ -365,6 +365,15 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
```
Comment (Laci21) [inline deploy/charts/istio-operator/templates/operator-rbac.yaml:368]: this might need to be added to the controller.go file as well to generate it for the kustomize manifests

File: internal/assets/manifests/istio-discovery/templates/telemetryv2_1.12.yaml [hunk 9]
```diff
@@ -14,6 +14,7 @@ metadata:
   labels:
```
Comment (Laci21) [inline internal/assets/manifests/istio-discovery/templates/telemetryv2_1.12.yaml:14]: I think it's fine to remove these 1.12 CRs this time.

File: internal/components/sidecarinjector/testdata/icp-expected-values.yaml [hunk 10]
```diff
@@ -1,6 +1,6 @@
-revision: cp-v113x
+revision: cp-v115x
 deployment:
-  image: banzaicloud/istio-sidecar-injector:v1.10.4-bzc.1
+  image: banzaicloud/istio-sidecar-injector:v1.15.0-bzc.1
```
Comment (Laci21) [inline internal/components/sidecarinjector/testdata/icp-expected-values.yaml:3]: Does this image actually exist?

File: internal/components/sidecarinjector/testdata/icp-test-cr.yaml [hunk 11]
```diff
@@ -30,7 +30,7 @@ spec:
         type: RollingUpdate
         rollingUpdate:
           maxUnavailable: 1
-      image: banzaicloud/istio-sidecar-injector:v1.10.4-bzc.1
+      image: banzaicloud/istio-sidecar-injector:v1.15.0-bzc.1
```
Comment (Laci21) [inline internal/components/sidecarinjector/testdata/icp-test-cr.yaml:33]: same Q here

File: internal/util/template.go [hunk 12]
```diff
@@ -106,6 +106,11 @@ func valueIfTemplateFunc(value interface{}) (string, error) {
 			return "", nil
 		}
 
+		m, ok := reflect.TypeOf(value).MethodByName("GetValue")
+		if ok {
+			value = m.Func.Call([]reflect.Value{reflect.ValueOf(value)})[0].Interface()
```
Comment (Laci21) [inline internal/util/template.go:111]: Could you pls explain to me what's happening here and why exactly? I'd like to learn :)

**Review by Laci21 — COMMENTED** [review]
Great work, @LuciferInLove! I've just added some nits/questions.

File: api/v1alpha1/istiocontrolplane.proto (hunk 3 above)
Comment (luciferinlove) [reply api/v1alpha1/istiocontrolplane.proto:128]: New buf doesn't support fields defined in multiple places. There was a recommendation to rename the fields in such a manner.

File: api/v1alpha1/istiocontrolplane_types.go (hunk 4 above)
Comment (luciferinlove) [reply api/v1alpha1/istiocontrolplane_types.go:57]: Because after moving to golang/protobuf this field contains mutex, so there were copylocks.

File: internal/components/sidecarinjector/testdata/icp-expected-values.yaml (hunk 10 above)
Comment (luciferinlove) [reply internal/components/sidecarinjector/testdata/icp-expected-values.yaml:3]: No, not yet

File: internal/util/template.go (hunk 12 above)
Comment (luciferinlove) [reply internal/util/template.go:111]: If a structure has `GetValue` method - call this method to get a value. This is needed because after moving to golang/protobuf there are wrappers for types:
https://github.com/golang/protobuf/blob/master/ptypes/wrappers/wrappers.pb.go

The same is for methods in `.tpl` files.

File: api/v1alpha1/common.proto (hunk 1 above)
Comment (Laci21) [reply api/v1alpha1/common.proto:63]: Talked about this offline, should be fine.

File: internal/components/sidecarinjector/testdata/icp-expected-values.yaml (hunk 10 above)
Comment (Laci21) [reply internal/components/sidecarinjector/testdata/icp-expected-values.yaml:3]: And will it? Seems like we haven't released a new one since `1.10.4`, will we have a new one for some reason?

File: internal/util/template.go (hunk 12 above)
Comment (Laci21) [reply internal/util/template.go:111]: Thanks!

**Review by Laci21 — APPROVED** [review]

File: internal/components/sidecarinjector/testdata/icp-expected-values.yaml (hunk 10 above)
Comment (luciferinlove) [reply internal/components/sidecarinjector/testdata/icp-expected-values.yaml:3]: This is only testdata, no real images will be downloaded.

Comment (Laci21) [reply internal/components/sidecarinjector/testdata/icp-expected-values.yaml:3]: oh, okay :)
